### PR TITLE
Add minimal savepoint support, to not fail

### PR DESCRIPTION
**Pull request by mauritsvanrees**
Fixes issue #35.

### CODE REVIEW

**Review by jamadden — COMMENTED** [review]
This scares me. It seems that you could ask for a savepoint, send some mails, roll the savepoint back, commit the transaction, and those mails would still get sent.

If I'm reading that correctly, that seems like very undesirable behaviour. I'd never want to see mails sent that had been rolled back. It seems that rather defeats the point of transactional mail delivery, doesn't it.

ISTM that if we can't actually support savepoints and rolling back, then an exception indicating that fact (fail fast) is better behaviour than silently ignoring desired transactional semantics.

**Comment by mauritsvanrees** [comment]
I confess to not completely knowing what I am doing when it comes to lower level transaction stuff, so thanks for having a look.

After your comment, I first started implementing a `rollback` method, but I saw this got called for datamanagers that were created *before* making a savepoint. In that case we must keep the mail.

And: it actually seems to work fine. I have updated the test to see what happens with a rollback.
Let me describe what the test now does, and what the transaction package does if I understand it correctly:

1. Transaction starts.
2. Send mail A. MailDataManager A holds delivery.
3. Create savepoint. So mail A *must* eventually be delivered, unless the complete transaction is aborted.
4. Send mail B. MailDataManager B holds delivery.
5. Do savepoint.rollback.
6. MailDataManager B is aborted, the MailDataManager is removed from the transaction.
7. For MailDataManager A `savepoint().rollback()` is called, but this does not have to do anything.
8. Send mail C. MailDataManager C holds delivery.
9. Commit the transaction.
10. MailDataManager A delivers mail A.
11. MailDataManager B was already aborted and basically does not exist anymore, so it delivers nothing.
12. MailDataManager C delivers mail C.
 
Does this seem safe to you after all?

**Comment by mauritsvanrees** [comment]
I do see a problem with the tests. Locally the `pypy` tests get flaky. I ran them 20 times, and 6 had failures, for example this:

```
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/threading.py", line 820, in __bootstrap_inner
    self.run()
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/threading.py", line 773, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/site-packages/zope/sendmail/tests/test_mailer.py", line 155, in run
    self.test_send_multiple_same_mailer()
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/site-packages/zope/sendmail/tests/test_mailer.py", line 139, in test_send_multiple_same_mailer
    self.mailer.send(fromaddr, toaddrs, msgtext)
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/site-packages/zope/sendmail/mailer.py", line 115, in send
    connection.sendmail(fromaddr, toaddrs, message)
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/site-packages/zope/sendmail/tests/test_mailer.py", line 42, in sendmail
    assert not self.closed
AssertionError



Failure in test test_send_multiple_threads (zope.sendmail.tests.test_mailer.TestSMTPMailerWithNoEHLO)
Traceback (most recent call last):
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/site-packages/zope/sendmail/tests/test_mailer.py", line 171, in test_send_multiple_threads
    self.assertEqual([True for _ in threads], results)
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/unittest/case.py", line 742, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/unittest/case.py", line 724, in assertSequenceEqual
    self.fail(msg)
  File "/Users/maurits/.pyenv/versions/pypy2.7-7.3.0/lib-python/2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: [True, True] != [True, AssertionError()]

First differing element 1:
True
AssertionError()

- [True, True]
+ [True, AssertionError()]

  Ran 81 tests with 1 failures, 0 errors and 3 skipped in 0.145 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
ERROR: InvocationError for command /Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/bin/zope-testrunner --test-path=src (exited with code 1)
```

Trying this on the master branch, all 20 go fine.

It does not go wrong directly in my new test, but it still may be caused by my changes.

PyPy3 looks fine.

**Comment by mauritsvanrees** [comment]
Ah, one in twenty PyPy3 test runs has a failure after all:

```
Failure in test test_send_multiple_threads (zope.sendmail.tests.test_mailer.TestSMTPMailer)
Traceback (most recent call last):
  File "/Users/maurits/.pyenv/versions/pypy3.6-7.3.0/lib-python/3/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/maurits/.pyenv/versions/pypy3.6-7.3.0/lib-python/3/unittest/case.py", line 605, in run
    testMethod()
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy3/site-packages/zope/sendmail/tests/test_mailer.py", line 171, in test_send_multiple_threads
    self.assertEqual([True for _ in threads], results)
  File "/Users/maurits/.pyenv/versions/pypy3.6-7.3.0/lib-python/3/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/maurits/.pyenv/versions/pypy3.6-7.3.0/lib-python/3/unittest/case.py", line 1028, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/Users/maurits/.pyenv/versions/pypy3.6-7.3.0/lib-python/3/unittest/case.py", line 1010, in assertSequenceEqual
    self.fail(msg)
  File "/Users/maurits/.pyenv/versions/pypy3.6-7.3.0/lib-python/3/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: [True, True] != [True, AssertionError("'me@example.com2' != 'me@[90 chars]n",)]

First differing element 1:
True
AssertionError("'me@example.com2' != 'me@[89 chars]\n",)

- [True, True]
+ [True,
+  AssertionError("'me@example.com2' != 'me@example.com1'\n- me@example.com2\n?               ^\n+ me@example.com1\n?               ^\n",)]
```

That looks like mails might sometimes be sent in a different order.

Are the `zope.sendmail` tests known to randomly fail?

File: src/zope/sendmail/tests/test_delivery.py
Comment (jamadden) [inline src/zope/sendmail/tests/test_delivery.py:286]: What do you think about adding a negative assertion here? There's the `len` test above, which *ought* to cover it, but I did miss that on the first read through. Or, to be even more explicit:
```python
subjects.sort()
self.assertEqual([b'Subject: one', b'Subject: three'], subjects)
```

**Review by jamadden — APPROVED** [review]
> I confess to not completely knowing what I am doing when it comes to lower level transaction stuff, so thanks for having a look.

And I confess to not looking at the greater context of how these particular objects interact with it!

> I have updated the test to see what happens with a rollback.

Thank you! That soothes my fears.

> Are the zope.sendmail tests known to randomly fail?

I think I recall the threading tests being flaky, yes.

File: src/zope/sendmail/tests/test_delivery.py
Comment (mauritsvanrees) [reply src/zope/sendmail/tests/test_delivery.py]: That may be clearer yes. Done.

**Comment by mauritsvanrees** [comment]
> I think I recall the threading tests being flaky, yes.

One failure I sometimes see, is this:

```
  File "/Users/maurits/community/plone-coredev/py3/src/zope.sendmail/.tox/pypy/site-packages/zope/sendmail/tests/test_mailer.py", line 42, in sendmail
    assert not self.closed
AssertionError
```

This is when running the thread tests from `test_send_multiple_threads`.
This may just mean that the mailer in the tests is not very thread-aware.

**Review by mgedmin — COMMENTED** [review]
The code looked scary, but the test soothed me.

I'd rather we wait for working CI (#37) before merging this one.

The tests are [known to be flaky (#15)](https://github.com/zopefoundation/zope.sendmail/issues/15). I never got to the bottom of this, so I don't know if it's a bug in the test, or an actual bug in the code.  Note that once the flakiness showed up on Python 2.7, so it's not a PyPy-specific issue.

**Comment by mauritsvanrees** [comment]
> The code looked scary, but the test soothed me.

Great!

> I'd rather we wait for working CI (#37) before merging this one.

The tests pass, but fine with me to wait.

> The tests are [known to be flaky (#15)](https://github.com/zopefoundation/zope.sendmail/issues/15). I never got to the bottom of this, so I don't know if it's a bug in the test, or an actual bug in the code. Note that once the flakiness showed up on Python 2.7, so it's not a PyPy-specific issue.

I hadn't looked for existing issue reports yet.
You have created a PR #38 to fix this. I am happy to wait for that to be approved and merged.

**Comment by mgedmin** [comment]
We have working CI and the flaky test has been fixed.  Would you mind rebasing on git master?

**Comment by mauritsvanrees** [comment]
Thanks for pushing through on this, @mgedmin !
I have rebased, and squashed into one commit for clarity.
Tox passes locally.

**Review by sallner — APPROVED** [review]
Looks good to me and ready to merge.

**Pull request merged**
